src/main/java/de/tum/in/www1/artemis/config/migration/entries/MigrationEntry20230808_203400.java

```diff
@@ -0,0 +1,460 @@
+     */
+    private void executeBatchedMigration(long start, long end, long exerciseCount, Map<ProgrammingExercise, Boolean> errorMap) {
+        // do batches
+        for (int batchStartIndex = (int) start, batch = 0; batchStartIndex < end; batchStartIndex++, batch = batch + BATCH_SIZE) {
+            var batchStart = System.currentTimeMillis();
+            Pageable pageable = PageRequest.of(batchStartIndex, BATCH_SIZE);
+
+            // TODO: in the unlikely case the buildPlanId is null, we could skip the template participation
+            var templateParticipationPage = templateProgrammingExerciseParticipationRepository.findAll(pageable);
+            log.info("Found {} template programming exercises to migrate in batch", templateParticipationPage.getTotalElements());
+            templateParticipationPage.map(templateParticipation -> {
+                if (!ciMigrationService.orElseThrow().buildPlanExists(templateParticipation.getProgrammingExercise().getProjectKey(),
+                        templateParticipation.getProgrammingExercise().getTemplateBuildPlanId())) {
+                    log.warn("Build plan {} does not exist for exercise {}", templateParticipation.getProgrammingExercise().getTemplateBuildPlanId(),
+                            templateParticipation.getProgrammingExercise().getId());
+                    log.info("Skipping migration for template of exercise {} and setting build_plan_id to null", templateParticipation.getProgrammingExercise().getId());
+                    templateParticipation.setBuildPlanId(null);
+                    templateProgrammingExerciseParticipationRepository.save(templateParticipation);
+                    return null;
+                }
+                log.info("Migrating build plan with name {} for exercise {}",
+                        templateParticipation.getProgrammingExercise().getProjectKey() + "-" + templateParticipation.getBuildPlanId(),
+                        templateParticipation.getProgrammingExercise().getId());
+                return templateParticipation;
+            }).filter(Objects::nonNull).forEach(templateParticipation -> {
+                var startMs = System.currentTimeMillis();
+                List<AuxiliaryRepository> auxiliaryRepositories = new ArrayList<>();
+                if (ciMigrationService.orElseThrow().supportsAuxiliaryRepositories()) {
+                    auxiliaryRepositories = auxiliaryRepositoryRepository.findByExerciseId(templateParticipation.getProgrammingExercise().getId());
+                }
+                migrateTemplateBuildPlan(templateParticipation.getProgrammingExercise(), auxiliaryRepositories, errorMap);
+
+                migrateTestBuildPlan(templateParticipation.getProgrammingExercise(), errorMap);
+                log.info("Migrated template build plan for exercise {} in {}ms", templateParticipation.getProgrammingExercise().getId(), System.currentTimeMillis() - startMs);
+            });
```

Comment:
I feel like this mapping and filtering would be better written linearly in this case so that it is easier to follow what is happening.

Reply:
Not too important as it's a migration